name_list Qt GUI

JeremyRand opened the pull request:
Port of the `name_list` Qt GUI to master branch.

Fixes https://github.com/namecoin/namecoin-core/issues/129

JeremyRand commented:
Handing off this PR to bonfire for the moment.

JeremyRand commented:
@brandonrobertz Why did your name tab branch have separate code paths for updating individual names, and for updating expirations?  Why not just use `name_list` each time?  Was `name_list` a performance bottleneck?

brandonrobertz commented:
As far as I remember, `name_list` didn't support all of the list operations required for displaying in the qt table model. I believe y'all (Daniel mostly) fixed the issues since.

JeremyRand commented:
> As far as I remember, `name_list` didn't support all of the list operations required for displaying in the qt table model. I believe y'all (Daniel mostly) fixed the issues since.

@brandonrobertz Okay great, so then I guess we'll nuke those extra code paths for individual names and expiration updates.  That makes the code a hell of a lot simpler, yay.  :)

JeremyRand commented:
@randy-waterhouse and @domob1812 , I think this is at the point where some cursory review from you would be helpful.

Things in particular that might especially merit testing are the interaction between this PR and either IBD or multiple wallets.

In terms of scope: this PR is deliberately of a limited scope.  Suggestions for legacy code that I can nuke are very welcome.  Reports of bugs that break existing functionality in master branch are very welcome.  Reports of bugs in the new functionality that can be fixed without increasing the size of the PR are very welcome.  However, I don't want to expand the scope of this PR with either additional features or bugfixes that involve significantly increasing the PR size.  Any additional features or bugfixes that need new code should be done as follow-up PR's, because I don't want this PR getting stuck in review hell.

Cheers.

brandonrobertz commented on src/qt/nametablemodel.cpp:
```diff
@@ -0,0 +1,319 @@
+    dataChanged(index(idx, 0), index(idx, columns.length()-1));
+}
+
+void
+NameTableModel::subscribeToCoreSignals()
+{
+    // Connect signals to wallet
+    //m_handler_transaction_changed = walletModel->wallet().handleTransactionChanged(std::bind(NotifyTransactionChanged, this, std::placeholders::_1, std::placeholders::_2));
+    //m_handler_show_progress = walletModel->wallet().handleShowProgress(std::bind(ShowProgress, this, std::placeholders::_1, std::placeholders::_2));
+}
+
+void
+NameTableModel::unsubscribeFromCoreSignals()
+{
+    // Disconnect signals from wallet
+    //m_handler_transaction_changed->disconnect();
+    //m_handler_show_progress->disconnect();
+}
```
Why not just wipe these out if we're not using this anymore? I think we should be good to go with the refreshing of the whole name table being triggered from `updateExpiration` and `processNewTransaction`.

JeremyRand replied:
Thanks, I've removed this (along with a bit of other dead code).

brandonrobertz reviewed (approved):
Overall I think this is good and fairly conservative in terms of changes.

JeremyRand commented:
> Things in particular that might especially merit testing are the interaction between this PR and either IBD or multiple wallets.

I tested IBD from scratch today, and this PR didn't seem to make a noticeable change in IBD speed.  Multiple wallets appear to work okay too, although I didn't test this very extensively.

JeremyRand commented:
@domob1812 Any chance you could take a look at this PR?  (If you think it's ready to merge, let me know and I'll squash the Git history before you merge it.)

randy-waterhouse commented:
I think there's issues with it. After building and running against current wallet/blockchain of the last version that supported names tabs (0.15.99-name-tab-beta1) it rewound blocks back 1 year then failed to launch the GUI with following error to CLI:

> ./namecoin-qt 
QStandardPaths: wrong ownership on runtime directory /run/user/1000, 1000 instead of 1002
>************************
>EXCEPTION: St13runtime_error       
>JSON value is not a string as expected       
>namecoin in Runaway exception       
>terminate called after throwing an instance of 'std::runtime_error'
>  what():  JSON value is not a string as expected
>Aborted (core dumped)

In the debug.log this seems to be the relevant bit:

>2020-10-19T02:28:37Z GUI: Platform customization: "other"
>2020-10-19T02:28:37Z GUI: Qt has caught an exception thrown from an event handler. Throwing
>exceptions from an event handler is not supported in Qt.
>You must not let any exception whatsoever propagate through Qt code.
>If that is not possible, in Qt 5 you must at least reimplement
>QCoreApplication::notify() and catch all exceptions there.
>
>2020-10-19T02:28:37Z 
>
>************************
>EXCEPTION: St13runtime_error       
>JSON value is not a string as expected       
>namecoin in Runaway exception

Grepping in src for "JSON value is not a string as expected " shows it is in univalue.h, which is called as include in "managenamespace.cpp" ... so I suspect something in there.

Futher testing by having no wallet.dat in .namecoin gets the GUI window open, rewinds and updates blocks from 1 year ago then fails in v. similar manner when selecting a wallet containing names:
>2020-10-19T03:23:44Z init message: Loading wallet...
>2020-10-19T03:23:48Z [wallet_bak.dat] Wallet File Version = 10500
>2020-10-19T03:23:48Z [wallet_bak.dat] Keys: 8664 plaintext, 0 encrypted, 3307 w/ metadata, 8664 total. Unknown wallet records: 0
>2020-10-19T03:23:48Z [wallet_bak.dat] Wallet completed loading in            3645ms
>2020-10-19T03:23:48Z init message: Rescanning...
>2020-10-19T03:23:48Z [wallet_bak.dat] Rescanning last 5 blocks (from block 529185)...
>2020-10-19T03:23:48Z [wallet_bak.dat] Rescan started from block >54dba79820b55ffdbc52a14e0900b95c3dc32d2dd7575e66dc723edb867f8cf2...
>2020-10-19T03:23:48Z [wallet_bak.dat] Rescan completed in               5ms
>2020-10-19T03:23:48Z name_list lookup error: Requested wallet does not exist or is not loaded
>2020-10-19T03:23:48Z [wallet_bak.dat] setKeyPool.size() = 1000
>2020-10-19T03:23:48Z [wallet_bak.dat] mapWallet.size() = 3418
>2020-10-19T03:23:48Z [wallet_bak.dat] m_address_book.size() = 38

From this test other debug.log fragments that maybe useful:

>2020-10-19T03:27:43Z [wallet_bak.dat] Submitting wtx ba05a3776b6eb015519e87553dcf85071060a166ff5cc23ab46f32378e53efc0 to mempool for relay
>2020-10-19T03:27:43Z [wallet_bak.dat] Submitting wtx f35a62f2d94c16227d5f1039ab04b3e1862d8cd55fded71d695dcfe4f63581d0 to mempool for relay
>2020-10-19T03:27:43Z GUI: Platform customization: "other"
>2020-10-19T03:27:44Z GUI: Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt.
>You must not let any exception whatsoever propagate through Qt code.
>If that is not possible, in Qt 5 you must at least reimplement
>QCoreApplication::notify() and catch all exceptions there.
>
>2020-10-19T03:27:44Z 
>
>************************
>EXCEPTION: St13runtime_error       
>JSON value is not a string as expected       
>namecoin in Runaway exception

JeremyRand commented:
Thanks for testing @randy-waterhouse .  Can you check whether anything similar happens in 4f110494a8d54fe71531f3a09261e0e19193bfc8 (i.e. the version of `master` that this branch is based on, but without this PR applied)?

JeremyRand commented:
@randy-waterhouse As an aside, do you have a current OpenPGP pubkey?

JeremyRand commented:
Marking this PR as WIP again until the issue from https://github.com/namecoin/namecoin-core/pull/376#issuecomment-711456174 is resolved.  AFAICT it accidentally happens to work in this PR without problems, but the code is still wrong, and I want to fix it before it's merged.

domob1812 commented on src/qt/bitcoingui.cpp:
```diff
@@ -277,6 +278,17 @@ void BitcoinGUI::createActions()
     tabGroup->addAction(historyAction);
 
+    manageNamesAction = new QAction(platformStyle->SingleColorIcon(":/icons/bitcoin_transparent_letter"), tr("&Manage Names"), this);
+    manageNamesAction->setStatusTip(tr("Manage names registered via Namecoin"));
```
What about changing the text to e.g. "Manage your registered names"?  I think that would be more descriptive (it is about the user's names not registered names in general, and they already know they are using Namecoin).

As a side effect, that would also avoid a rebranding for Xaya (but that's of course not a reason to influence anything for Namecoin).

I'm fine with anything you decide, though.

JeremyRand replied:
"Your" is inconsistent with existing text, e.g. "Browse transaction history", which also implicitly refers to the wallet as opposed to the entire set of transactions in the chain.  No objection to "Manage registered names" though.

domob1812 commented on src/qt/forms/managenamespage.ui:
```diff
@@ -0,0 +1,253 @@
+      <item>
+       <widget class="QValidatedLineEdit" name="registerName">
+        <property name="toolTip">
+         <string>Enter a name or domain name (prefixed with d/) to be registered via Namecoin.</string>
```
Perhaps here it could say "...to be registered on the blockchain".  Again totally up to you.

JeremyRand replied:
How about "to register it"?  Avoids the awkward passive voice, and avoids the presumably obvious statement that the registration is based on a blockchain.

domob1812 commented on src/qt/managenamespage.h:
```diff
@@ -0,0 +1,56 @@
+
+private:
+    const PlatformStyle *platformStyle;
+    Ui::ManageNamesPage *ui;
```
I think we had a discussion about this previously, and it may just be how things are done with Qt (which I don't know anything about).  But in general, I think it would be better to use `std::unique_ptr` here instead of a raw pointer.

JeremyRand replied:
I just ran `grep "Ui::" ./src/qt/*.h`, and it indicates that a raw pointer is consistent with upstream, so I prefer to keep it like this.

domob1812 replied:
Yes, I think most of the code does that, so it may be a general Qt pattern (or it may just be legacy that has not been corrected yet).  I'm definitely fine with following the existing Qt code.

domob1812 commented on src/qt/nametablemodel.cpp:
```diff
@@ -0,0 +1,283 @@
+class NameTablePriv
+{
+public:
+    explicit NameTablePriv(NameTableModel *_parent) :
```
If `_parent` is never null (not sure, but the code sometimes just dereferences it without checking), then perhaps this could be made a reference instead.

JeremyRand replied:
Indeed, it is never null.  I'll make it a reference.

domob1812 commented on src/qt/walletmodel.cpp:
```diff
@@ -28,13 +28,25 @@
 #include <QTimer>
 
+// TODO: figure out which of these includes are actually still necessary for name_list
+#include <boost/foreach.hpp>
```
At least the boost include should not be necessary I think.  The codebase has stopped using boost pretty much everywhere for a long time, and we should definitely use C++11 for-each loops instead of boost.

JeremyRand replied:
You are correct, the boost include wasn't needed anymore.

domob1812 reviewed (approved):
Looks good as far as I can tell without diving too deeply into the Qt bits.  I've just commented on a few nits.

JeremyRand commented:
@randy-waterhouse I can reproduce your crash now.  It happens when a name or value has invalid encoding according to the configured encoding rules.  Should be easy to fix.

randy-waterhouse commented:
> @randy-waterhouse I can reproduce your crash now. It happens when a name or value has invalid encoding according to the configured encoding rules. Should be easy to fix.

That's good.

Also I was wondering if there is a plan to reproduce any other features of Brandon's "Manage Names" tab. E.g. the "New name:" form line, "Submit" button, "Configure name", "Renew name" buttons?

JeremyRand commented:
@domob1812 @randy-waterhouse Okay, I think I've addressed all comments so far.  Let me know if this is ready to squash or if further changes are needed.  (I haven't squashed yet so that it's easier to compare to my previous iteration, I'll squash it before a merge though.)

The fix for Randy's crash works around it by not displaying any names whose name/value has invalid encoding.  So that probably means that, in the wallet you tested with, at least 1 name will be missing from the table.  It should be possible to have smarter handling of this case (i.e. retrieving the name/value as hex), but I'd prefer to do that as a follow-up PR.

JeremyRand commented:
> Also I was wondering if there is a plan to reproduce any other features of Brandon's "Manage Names" tab. E.g. the "New name:" form line, "Submit" button, "Configure name", "Renew name" buttons?

@randy-waterhouse Yes, I intend to get those added as follow-up PR's.  The "Renew name" functionality is already mostly implemented in https://github.com/namecoin/namecoin-core/pull/376 ; I'll ping you from any follow-up PR's so you can review them when they're ready.

JeremyRand commented:
Regarding review from @randy-waterhouse at https://github.com/namecoin/namecoin-core/pull/376#issuecomment-719067415 :

> I've tested now and the crash on opening old wallet is fixed.

Great!

> expired names associated with keys in the wallet display in the Manage Names tab list, with a negative XX-XX-XX expiry date

This is due to https://github.com/namecoin/namecoin-core/issues/379 .  Also see https://github.com/namecoin/namecoin-core/issues/380 .  I'd prefer to fix those at the RPC layer, and then tweak the RPC arguments that the Qt GUI uses; these would be considered out of scope for this PR.

> in the Transactions tab "Label" column there is no information displayed about the name or name operation associated with a name operation transaction, only the public keys involved

This is also the case for the label displayed in the RPC interface, right?  If so, I think we should fix it properly (i.e. the code for setting/getting the label, which is agnostic to Qt/RPC) rather than messing with the Qt-specific display code.

JeremyRand commented:
@domob1812 @randy-waterhouse Are there any remaining issues that I need to solve here before I squash it in preparation for a merge?

domob1812 commented on src/qt/managenamespage.cpp:
```diff
@@ -0,0 +1,143 @@
+    if (event->type() == QEvent::FocusIn)
+    {
+        if (object == ui->tableView)
+        {
```
Should there be something in here?  If not, then you should remove the two `if` as well here (and perhaps add them in later together with a change that actually puts conditional code here).

JeremyRand replied:
This `if` statement is going to contain stuff for the the Submit Name and Configure Name buttons, which aren't currently present in this PR but will be added later.  I can factor out the `if` into a separate commit and move that commit into a separate PR.

JeremyRand replied:
Moved this code to https://github.com/namecoin/namecoin-core/pull/376

domob1812 commented on src/qt/managenamespage.cpp:
```diff
@@ -0,0 +1,143 @@
+
+    const bool state = table->selectionModel()->hasSelection();
+    //ui->configureNameButton->setEnabled(state);
+    //ui->renewNameButton->setEnabled(state);
```
I think you should remove the commented out lines here, and just add them in with the future changes that allow renewing names.  This is not directly related to this PR.  (But if you really prefer to keep them in, that's also fine with me.)

JeremyRand replied:
Moved this code to https://github.com/namecoin/namecoin-core/pull/376

domob1812 reviewed (approved):
Two nits, but otherwise all is fine from my point of view.

randy-waterhouse commented:
> @domob1812 @randy-waterhouse Are there any remaining issues that I need to solve here before I squash it in preparation for a merge?

No issues that I've seen recently.

JeremyRand commented:
@domob1812 Okay, I've addressed both nits and squashed.  Feel free to merge if you think it's ready.

domob1812 commented:
ACK b6d5be95b871333877739bfd0ad26be5cdccf4c7.

domob1812 commented:
Merged.  Very cool!
